fastly/user.go

```diff
@@ -0,0 +1,193 @@
+	DeletedAt              *time.Time `mapstructure:"deleted_at"`
+}
+
+// usersByLogin is a sortable list of users.
+type usersByName []*User
+
+// Len, Swap, and Less implement the sortable interface.
+func (s usersByName) Len() int      { return len(s) }
+func (s usersByName) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s usersByName) Less(i, j int) bool {
+	return s[i].Name < s[j].Name
+}
+
+// ListUsers returns the full list of users belonging to the currently
+// authenticated user.
+func (c *Client) ListUsers() ([]*User, error) {
+	resp, err := c.Get("/users", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var t []*User
+	if err := decodeJSON(&t, resp.Body); err != nil {
+		return nil, err
+	}
+	sort.Stable(usersByName(t))
+	return t, nil
+}
+
+// ListCustomerUsersInput is used as input to the ListCustomerUsers function.
+type ListCustomerUsersInput struct {
+	ID string
+}
+
+// ListCustomerUsers returns the full list of users belonging to a specific
+// customer.
+func (c *Client) ListCustomerUsers(i *ListCustomerUsersInput) ([]*User, error) {
+	if i.ID == "" {
+		return nil, ErrMissingID
+	}
+
+	path := fmt.Sprintf("/customer/%s/users", i.ID)
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var u []*User
+	if err := decodeJSON(&u, resp.Body); err != nil {
+		return nil, err
+	}
+	sort.Stable(usersByName(u))
+	return u, nil
+}
+
+// GetCurrentUser retrieves the user information for the authenticated user.
+func (c *Client) GetCurrentUser() (*User, error) {
+	resp, err := c.Get("/current_user", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var u *User
+	if err := decodeJSON(&u, resp.Body); err != nil {
+		return nil, err
+	}
+
+	return u, nil
+}
+
+// GetUserInput is used as input to the GetUser function.
+type GetUserInput struct {
+	ID string
+}
+
+// GetUser retrieves the user information for the user with the given
+// id. If no user exists for the given id, the API returns a 404 response.
+func (c *Client) GetUser(i *GetUserInput) (*User, error) {
+	if i.ID == "" {
+		return nil, ErrMissingID
+	}
+
+	path := fmt.Sprintf("/user/%s", i.ID)
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var u *User
+	if err := decodeJSON(&u, resp.Body); err != nil {
+		return nil, err
+	}
+
+	return u, nil
+}
+
+// CreateUserInput is used as input to the CreateUser function.
+type CreateUserInput struct {
+	Login string `form:"login"`
+	Name  string `form:"name,omitempty"`
+	Role  string `form:"role,omitempty"`
+}
+
+// CreateUser creates a new API token with the given information.
+func (c *Client) CreateUser(i *CreateUserInput) (*User, error) {
+	resp, err := c.PostForm("/user", i, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var u *User
+	if err := decodeJSON(&u, resp.Body); err != nil {
+		return nil, err
+	}
+	return u, nil
+}
+
+// UpdateUserInput is used as input to the UpdateUser function.
+type UpdateUserInput struct {
+	ID string `form:"-"`
```

Comment:
``form:"-"`` is needed, so that it is not serialized into the update request...
otherwise we get `422 - Unprocessable Entity` with...
```
            Title:  Invalid attribute
            Detail: unknown attribute 'ID' for User.
```

Reply:
Yes, you're absolutely right about this one!

```
2020/02/25 23:52:43 -->
	Method:	PUT
	URL:	https://api.fastly.com/user/44pSHnkJmZicERXxOtAOAc
	Body:	{%!s(*strings.Reader=&{ID=44pSHnkJmZicERXxOtAOAc&name=Shopify 0 -1})}
	Fastly-Key, my_key
	User-Agent, FastlyGo/1.5.0 (+github.com/fastly/go-fastly; go1.12)
	Content-Type, application/x-www-form-urlencoded
2020/02/25 23:52:44 <--
	Status:	422
2020/02/25 23:52:44 422 - Unprocessable Entity:

    Title:  Invalid attribute
    Detail: unknown attribute 'ID' for User.
exit status 1
```